cmd/server-mux.go

```diff
@@ -420,21 +451,113 @@ func (m *ServerMux) ListenAndServe(certFile, keyFile string) (err error) {
 		MinVersion: tls.VersionTLS12,
 	} // Always instantiate.
 
-	if tlsEnabled {
-		// Configure TLS in the server
-		if config.NextProtos == nil {
-			config.NextProtos = []string{"http/1.1", "h2"}
+	// Configure TLS in the server
+	if config.NextProtos == nil {
+		config.NextProtos = []string{"http/1.1", "h2"}
+	}
+	config.Certificates = make([]tls.Certificate, 1)
+
+	config.Certificates[0], tlsErr = tls.X509KeyPair(m.certPEM, m.keyPEM)
+	if tlsErr != nil {
+		return nil, false, false, tlsErr
+	}
+
+	// Fetch for OCSP response
+	ocspBytes, _, ocspErr := ocsp.GetOCSPForCert(m.certPEM)
+	if ocspErr == nil {
+		ocspUpdated = true
+		config.Certificates[0].OCSPStaple = ocspBytes
+	} else {
+		if strings.Contains(ocspErr.Error(), "no OCSP server specified in cert") {
+			ocspUnsupported = true
 		}
-		config.Certificates = make([]tls.Certificate, 1)
-		config.Certificates[0], err = tls.LoadX509KeyPair(certFile, keyFile)
-		if err != nil {
-			return err
+	}
+
+	// Save the build tls config
+	m.tlsMu.Lock()
+	m.tlsConfig = config
+	m.tlsMu.Unlock()
+
+	return config, ocspUpdated, ocspUnsupported, nil
+}
+
+// ocspSuccessUpdateInterval is the interval to the next
+// ocsp update when the current one was successful
+const ocspSuccessUpdateInterval = time.Hour
+
+// ocspFailUpdateInterval is the interval to the next
+// ocsp update when the current one was failed.
+const ocspFailUpdateInterval = time.Minute
+
+// tlsBackgroundUpdate - regularly builds tls config to update
+// ocsp response. A totally new tls.Config will be built to avoid
+// any race accessing to the tls config structure between all
+// goroutines.
+func (m *ServerMux) tlsBackgroundUpdate() {
+	go func() {
+		// Infinite loop to update OCSP value
+		for {
+			// Build new TLS config, with updated ocsp
+			_, ocspUpdated, ocspUnsupported, _ := m.buildSrvTLSConfig()
+			// Certificate doesn't support ocsp extension
+			if ocspUnsupported {
+				// Wait until we get order to finish
+				<-m.tlsUpdateDoneCh
```

Comment:
Why do we need to read from `m.tlsUpdateDoneCh` before returning?  This will make long wait before returning.